File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& a, int k) {
        int n = a.size();
        vector<int> p(n,0);
        for(int i =0; i < n; i++)
            p[i] = p[(i-1)>=0?i-1:i]+a[i];
        map<int,int> m;
        m[0]=1;
        int ans = 0;
        for(int  i = 0; i < n; i++){
            if(m[p[i]-k]){
                ans+=m[p[i]-k];
            }
            m[p[i]]++;
        }
        return ans;
    }
};

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, a: List[int], t: int) -> int:
        ss = 0
        d = defaultdict(int)
        d[0]=1
        n = len(a)
        pref = [0]*n
        pref[0]=a[0]
        ct = 0
        for j in range(1,n):
            pref[j]=a[j]+pref[j-1]
        for r in range(n):
            x = pref[r]-t
            if x in d:
                ct+=d[x]
            d[pref[r]]+=1


        return ct
        